pkg/cloudprovider/provider/cherryservers/provider.go

```diff
@@ -0,0 +1,358 @@
+	}
+
+	if c.ProjectID != "" {
+		if _, _, err := client.Project.List(c.ProjectID); err != nil {
+			return fmt.Errorf("failed to get projectID")
+		}
+	}
+
+	return nil
+}
+
+func (p *provider) Create(machine *v1alpha1.Machine, _ *cloudprovidertypes.ProviderData, userdata string) (instance.Instance, error) {
+	c, pc, err := p.getConfig(machine.Spec.ProviderSpec)
+	if err != nil {
+		return nil, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Failed to parse MachineSpec, due to %v", err),
+		}
+	}
+
+	client := getClient(c.Token)
+
+	osName, err := getNameForOS(pc.OperatingSystem)
+	if err != nil {
+		return nil, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Invalid operating system specified %q, details = %v", pc.OperatingSystem, err),
+		}
+	}
+
+	teamID, err := strconv.Atoi(c.TeamID)
+	if err != nil {
+		return nil, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Invalid TeamID, details = %v", err),
+		}
+	}
+
+	var planID int
+	plans, _, err := client.Plans.List(teamID)
+	for _, plan := range plans {
+		if plan.Name == c.Plan {
+			planID = plan.ID
+			break
+		}
+	}
+
+	serverCreateRequest := cherrygo.CreateServer{
+		ProjectID:   c.ProjectID,
+		Hostname:    machine.Name,
+		Image:       osName,
+		Region:      c.Location,
+		SSHKeys:     []string{},
+		IPAddresses: []string{},
+		PlanID:      strconv.Itoa(planID),
+		UserData:    base64.StdEncoding.EncodeToString([]byte(userdata)),
+	}
+
+	server, _, err := client.Server.Create(c.ProjectID, &serverCreateRequest)
+	if err != nil {
+		return nil, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Failed to create server = %v", err),
+		}
+	}
+
+	return &cherryServer{server: server}, nil
+}
+
+type cherryServer struct {
+	server cherrygo.Server
+}
+
+func (s *cherryServer) Name() string {
+	return s.server.Hostname
+}
+
+func (s *cherryServer) ID() string {
+	return strconv.Itoa(s.server.ID)
+}
+
+func (s *cherryServer) Addresses() []string {
+	var addresses []string
+	for _, ip := range s.server.IPAddresses {
+		addresses = append(addresses, ip.Address)
+	}
+	return addresses
+}
+
+func (s *cherryServer) Status() instance.Status {
+	if s.server.State != "active" {
+		return instance.StatusCreating
+	}
+	return instance.StatusRunning
+}
+
+func (p *provider) Get(machine *v1alpha1.Machine, _ *cloudprovidertypes.ProviderData) (instance.Instance, error) {
+	c, _, err := p.getConfig(machine.Spec.ProviderSpec)
+	if err != nil {
+		return nil, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Failed to parse MachineSpec, due to %v", err),
+		}
+	}
+
+	client := getClient(c.Token)
+
+	servers, _, err := client.Servers.List(c.ProjectID)
+	if err != nil {
+		return nil, err
```

Comment:
Should we return TerminalError here too?

Reply:
TerminalError retruned